ompi/request/request.h

```diff
@@ -457,6 +459,21 @@ static inline int ompi_request_complete(ompi_request_t* request, bool with_signa
     return OMPI_SUCCESS;
 }
 
+static inline int ompi_request_set_callback(ompi_request_t* request,
+                                            ompi_request_complete_fn_t cb,
+                                            void* cb_data)
+{
+    request->req_complete_cb_data = cb_data;
+    request->req_complete_cb = cb;
+    /* If request is completed and the callback is not called, need to call callback */
+    if ((NULL != request->req_complete_cb) && (request->req_complete == REQUEST_COMPLETED)) {
```

Comment:
I believe there is a subtle race here with the code in `ompi_request_complete` (that probably has existed before). If   `ompi_request_set_callback` is executed while another thread is in between the check for `req_complete_cb` and the CAS on `req_complete` (around [here](https://github.com/open-mpi/ompi/pull/7716/files#diff-a134394a525737f997514a301e7682fbR443)) then a) the current thread won't see that the request is completed; and b) the thread in `ompi_request_complete` won't see that a callback has been registered. I guess the testing/setting of `req_complete_cb` has to be done using CAS and some magic value indicating that the request is complete.

Reply:
Yes, this is a known pb we had in ages (as soon as you enable the library in threaded mode). All potential solutions are too costly, and the window of opportunity is too narrow.